### PR TITLE
Memberships & Subscriptions: add logout API

### PR DESCRIPTION
**Pull request by simison**
## Proposed changes:

* Adds a way to "logout" as a member from the site; basically just clearing out the membership cookie.
* Accepts both get/post requests; allows flexibility to not leave trace in browser history by using `post`, but also flexibility we need with simple `get` when calling it from Calypso.

Next step is to add a simple "logout" link in Calypso to subscription management page: 

Screenshot 2024-01-12 at 14 29 40

In future potentially "logout" functionality in ["Subscriber login" block](https://github.com/Automattic/jetpack/issues/34884) as well. cc @pkuliga 

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

* On any page at blog where you're subscribed to, append `?jp-memberships-logout` to the URL and note how cookie gets cleared and you get logged out from subscriber-only content.

### CODE REVIEW

File: projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-abstract-token-subscription-service.php [hunk 1]
```diff
@@ -36,6 +37,20 @@ abstract class Abstract_Token_Subscription_Service implements Subscription_Servi
 	 */
 	public function initialize() {
 		$this->get_and_set_token_from_request();
+		$this->do_logout_from_request();
```
Comment (simison) [inline projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-abstract-token-subscription-service.php:40]: Wasn't sure if this is the best place to handle this; open for feedback.

File: projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-abstract-token-subscription-service.php [hunk 2]
```diff
@@ -448,6 +469,22 @@ private function set_token_cookie( $token ) {
...
+	 * @return void
+	 */
+	private function clear_token_cookie() {
+		if ( defined( 'TESTING_IN_JETPACK' ) && TESTING_IN_JETPACK ) {
+			return;
+		}
+
+		if ( isset( $_COOKIE[ self::JWT_AUTH_TOKEN_COOKIE_NAME ] ) && false === headers_sent() ) {
+			unset( $_COOKIE[ self::JWT_AUTH_TOKEN_COOKIE_NAME ] ); // In case cookie gets checked later in code
+			setcookie( self::JWT_AUTH_TOKEN_COOKIE_NAME, '', 1, '/', COOKIE_DOMAIN, is_ssl(), true ); // httponly -- used by visitor_can_view_content() within the PHP context.
```
Comment (chrisfromthelc) [inline projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-abstract-token-subscription-service.php:484]: I'm not seeing this work as expected on simple sites with a *.wordpress.com address. It appears `COOKIE_DOMAIN` is detecting `.wordpress.com`, but the cookie domain in my Simple test site is `.chriswsmith9.wordpress.com`.

Comment (pkuliga) [reply projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-abstract-token-subscription-service.php:484]: It doesn't work for me for the same reason.

**Comment by n3f** [comment]
FYI some previous work was in #33093. (It closed today automatically after no changes).  Also linking to https://github.com/Automattic/gold/issues/126.

**Comment by simison** [comment]
@n3f cool! Thanks for sharing.

We just chatted about similar approach with @pkuliga as highlighted in https://github.com/Automattic/gold/issues/126#issuecomment-1719304238

**Comment by pkuliga** [comment]
@simison I moved some logic to https://github.com/Automattic/jetpack/pull/34981 . Do you think we can close this or you still want to work on this?

**Comment by simison** [comment]
I think we might still need the "?p-memberships-logout' URL handling from this PR to be able to link to logout from subscription management page, so just a rebase.

File: projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-abstract-token-subscription-service.php (hunk 2 above)
Comment (millerf) [reply projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-abstract-token-subscription-service.php:484]: We do not rely on the cookie on `*.wordpress.com` address...

**Pull request closed without merging**